x/btcstaking/keeper/power_dist_change.go

```diff
@@ -196,6 +197,11 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 
 		// add this finality provider to the new cache if it has voting power
 		if fp.TotalVotingPower > 0 {
+			// voting power is not assigned if it does not have timestamped
+			// public randomness for this height
+			if !k.FinalityKeeper.HasTimestampedPubRand(ctx, fp.BtcPk, height) {
```

Comment:
hmm with this addition `ProcessAllPowerDistUpdateEvents` is probably in dire need of optimization, as now for every finality provider we iterate over potentially all pub rand commits of this finality provider. 

This can be even used as attack vector is we do not set `MinPubRand` high enough i.e create a lot if finality providers with minimal amount of delegation and super small random commits, to make this method run a lot of time.

Reply:
True. But not sure it has a real impact because this method is only called when the fp has voting power. We can revisit potential optimization when we do the refactoring. Added a TODO for now